File: data_preprocessing/event_streamer.py
from collections import namedtuple
import struct
import numpy as np

# Return type
Event = namedtuple("Event", "x y polarity timestamp")

# EVT2 events
CD_ON = 0x1
CD_OFF = 0x0
EVT_TIME_HIGH = 0x8


class EventStream:

    data = []
    fpath = ""

    event_buffer = []

    # Bit masks for extracting event info
    mask_6b = np.uint32(0x3F)
    mask_11b = np.uint32(0x7FF)
    mask_28b = np.uint32(0xFFFFFFF)

    # Start at byte 239 to skip the EVT2 file headers
    last_read_byte = 239

    time_high = np.uint64(0)

    def __init__(self, fpath):
        self.fpath = fpath

    def read(self) -> Event:
        if len(self.event_buffer) > 0:
            return self.event_buffer.pop(0)
        else:
            with open(self.fpath, "rb") as f:
                f.seek(self.last_read_byte)

                while len(self.event_buffer) < 1000:
                    byte_buffer = f.read(4096)
                    if not byte_buffer:
                        self.event_buffer.append(None)
                        return

                    for data in struct.iter_unpack("I", byte_buffer):
                        self.last_read_byte += 4
                        data = data[0]

                        event_type = data >> 28
                        if event_type <= 0x1:
                            # Combine lower half with upper half of timestamp
                            timestamp = self.time_high << 6 | ((data >> 22) & self.mask_6b)

                            event_x = data >> 11 & self.mask_11b
                            event_y = data & self.mask_11b

                            polarity = event_type

                            self.event_buffer.append(Event(event_x, event_y, polarity, timestamp))

                        elif (data >> 28) == 0x8:
                            # Extract upper half of full timestamp
                            self.time_high = data & self.mask_28b

            return self.read()

File: data_preprocessing/event_streamer.c
#include <stdbool.h>
#include <stdio.h>

#define CHUNK_SIZE 4096

enum EVENT_TYPE {
    CD_OFF = 0x0,
    CD_ON = 0x1,
    EVT_TIME_HIGH = 0x8,
};

struct Event {
    int x;
    int y;
    bool polarity;
    long timestamp;
} typedef Event;

unsigned int mask_6b = 0x3F;
unsigned int mask_11b = 0x7FF;
unsigned int mask_28b = 0xFFFFFFF;

void read_window(long *read_from, long *time_high, Event *event_buffer,
                 int event_buffer_size) {

    FILE *file_handle = fopen("events.raw", "rb");

    unsigned char buffer[CHUNK_SIZE];

    *time_high = 0;
    long window_start = 0;
    long timestamp = 0;
    int event_x, event_y;

    int event_idx = 0;

    fseek(file_handle, *read_from, 0);
    while (fread(buffer, sizeof(buffer), 1, file_handle) &&
           event_idx < event_buffer_size) {
        for (size_t i = 0; i < CHUNK_SIZE; i += 4) {
            if (event_idx >= event_buffer_size) {
                break;
            }
            *read_from += 4;
            unsigned int data = 0;
            data = buffer[i] | buffer[i + 1] << 8 | buffer[i + 2] << 16 |
                   buffer[i + 3] << 24;

            unsigned char event_type = data >> 28;

            if (event_type == CD_OFF || event_type == CD_ON) {
                // Combine lower half with upper half of timestamp
                timestamp = *time_high << 6 | ((data >> 22) & mask_6b);
                // if (timestamp < 1000) {
                //     continue;
                // } //  else if (window_start <= 0 && timestamp > 1000) {
                //     window_start = timestamp;
                // } else if (window_start > 0 &&
                //            (timestamp - window_start) > window_len) {
                //     break;
                // }

                event_x = data >> 11 & mask_11b;
                event_y = data & mask_11b;

                Event curr_event = {event_x, event_y, event_type, timestamp};
                event_buffer[event_idx++] = curr_event;

                // if (event_type) {
                //     printf("CD_ON\n");
                // } else {
                //     printf("CD_OFF\n");
                // }

                // printf("X: %d\n", event_x);
                // printf("Y: %d\n", event_y);
                // printf("timestamp: %ld\n", timestamp);
            } else if (event_type == EVT_TIME_HIGH) {
                // printf("EVT_TIME_HIGH\n");
                //  Extract upper half of full timestamp
                *time_high = data & mask_28b;
            }
        }
    }

    fclose(file_handle);
}

// int main(int argc, char const *argv[]) {
//     int a = 239;
//     long b = 0;
//     read_window(&a, 10000, &b, NULL);
//     return 0;
// }
